### PR TITLE
Adjust padding in batch generation

### PR DESCRIPTION
**Pull request by gaetanlop**
# What does this PR do?

Fix a bug in the `batch_generation` function where we try to concatenate tensors that can be of different dimensions (`sequence_length`) when generating with `mini_batches`.

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
- [X] Did you write any new necessary tests?

## Who can review?

@kashif @qgallouedec

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -250,3 +251,60 @@ def test_data_collator_for_chatml(self):
...
+
+        self.assertGreater(max_length, context_length)
+        self.assertEqual(query_responses.shape, (bs, max_length))
+        self.assertEqual(logits.shape, (bs, max_length, self.model.config.vocab_size))
+
+    def test_single_batch_generation(self):
+        batch = [
+            self.tokenizer.apply_chat_template(example[:-1], add_generation_prompt=True, tokenize=False)
+            for example in self.examples
+        ]
```
Comment (kashif) [inline tests/test_utils.py:298]:
```suggestion
       single_example = self.examples[0]
        batch = [
            self.tokenizer.apply_chat_template(single_example[:-1], add_generation_prompt=True, tokenize=False)
        ]
```

**Comment by kashif** [comment]
@gaetanlop  can we use the `pad` helpers?
```
padded_query_responses = pad(query_responses, padding_value=pad_token_id)
padded_logitss = pad(logitss, padding_value=float('-inf'))
```

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -1284,6 +1284,15 @@ def batch_generation(
...
         logitss.append(logits)
+
+    # padding tensors
+    max_length = max(query_response.shape[1] for query_response in query_responses)
+    query_responses = [
+        torch.nn.functional.pad(query_response, (0, max_length - query_response.shape[1]), value=pad_token_id)
+        for query_response in query_responses
+    ]
+    logitss = [torch.nn.functional.pad(logits, (0, 0, 0, max_length - logits.shape[1]), value=0) for logits in logitss]
+
```
Comment (kashif) [inline trl/trainer/utils.py:1295]:
```suggestion
    # padding tensors
    padded_query_responses = pad(query_responses, padding_value=pad_token_id, padding_side="right")
    padded_logitss = pad(logitss, padding_value=0, padding_side="right")

```

**Comment by gaetanlop** [comment]
@kashif, the `pad` function expects the tensor to have no leading dimension corresponding to the batch size.

Here is an example `query_responses`:
```python
query_responses = [
  torch.randint(vocab_size, (bs, seq_length1)), 
  torch.randint(vocab_size, (bs, seq_length2)), 
  torch.randint(vocab_size, (remaining_samples, seq_length3))
]
```

Using the `pad` function as it is would require the following change before passing the `query_responses` to the `pad` function: 

```python
query_responses=[query_reps[i] for query_reps in query_responses for i in range(query_reps.size(0))]
```

We can also change the pad function? What do you prefer?

**Review by kashif — APPROVED** [review]

File: trl/trainer/utils.py [hunk 3]
```diff
@@ -1284,6 +1284,11 @@ def batch_generation(
         )
         query_responses.append(query_response)
         logitss.append(logits)
+
+    # padding tensors
+    padded_query_responses = pad(query_responses, padding_value=pad_token_id, padding_side="right")
+    padded_logitss = pad(logitss, padding_value=0, padding_side="right")
+
     return torch.cat(query_responses, 0), torch.cat(logitss, 0)
```
Comment (kashif) [inline trl/trainer/utils.py:1292]:
```suggestion
    return padded_query_responses, padded_logitss
```

**Comment by gaetanlop** [comment]
This won't work @kashif, it still requires reshaping the tensors

**Comment by kashif** [comment]
ah damn! my bad sorry!

**Comment by gaetanlop** [comment]
This should be fixed now
